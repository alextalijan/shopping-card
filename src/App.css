.header {
  background-color: #1e293b;
  color: white;
  height: 100px;
  display: flex;
  justify-content: end;
  align-items: center;
  gap: 2rem;
}

.nav-link {
  text-decoration: none;
  font-size: 2rem;
  transition: transform 0.3s;
}

.nav-link:hover {
  transform: scale(1.2);
}

.nav-link:visited {
  color: white;
}

.nav-link:last-child {
  margin-right: 2rem;
}
