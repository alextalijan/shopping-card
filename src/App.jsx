import { useEffect, useState } from 'react';
import './App.css';
import { Outlet, Link } from 'react-router-dom';

function App() {
  const [cart, setCart] = useState([]);
  const [shopItems, setShopItems] = useState([]);
  const [loadingShop, setLoadingShop] = useState(true);
  const [shopError, setShopError] = useState(null);

  function addToCart(id, title, imageSrc, price, amount) {
    if (amount === 0) return;

    const addedItem = {
      id,
      title,
      imageSrc,
      price,
      amount,
    };

    for (const item of cart) {
      if (addedItem.id === item.id) {
        const newCart = cart.map((item) => {
          if (item.id === addedItem.id) {
            return {
              id: addedItem.id,
              title: addedItem.title,
              price: addedItem.price,
              amount: item.amount + addedItem.amount,
            };
          }

          return item;
        });

        setCart(newCart);
        return;
      }
    }

    setCart([...cart, addedItem]);
  }

  function removeFromCart(id) {
    setCart(cart.filter((item) => item.id !== id));
  }

  useEffect(() => {
    fetch('https://fakestoreapi.com/products/')
      .then((response) => {
        if (!response.ok) {
          throw new Error(
            'Failed to load items from the shop. Please try again by refreshing the page.',
          );
        }

        return response.json();
      })
      .then((data) => {
        const items = [];

        for (let i = 0; i < 10; i += 1) {
          items.push({
            id: data[i].id,
            title: data[i].title,
            price: data[i].price,
            imageSrc: data[i].image,
          });
        }

        setShopItems(items);
      })
      .catch((error) => {
        setShopError(error.message);
      })
      .finally(() => {
        setLoadingShop(false);
      });
  }, []);

  return (
    <>
      <header className="header">
        <Link className="nav-link" to="/">
          Home
        </Link>
        <Link className="nav-link" to="/shop">
          Shop
        </Link>
        <Link className="nav-link" to="/cart">
          Cart <span>({cart.length})</span>
        </Link>
      </header>
      <main>
        <Outlet
          context={{
            shopItems,
            loadingShop,
            shopError,
            addToCart,
            cart,
            setCart,
            removeFromCart,
          }}
        />
      </main>
    </>
  );
}

export default App;
