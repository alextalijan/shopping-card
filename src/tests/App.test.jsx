import { describe, it, expect } from 'vitest';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { screen, render } from '@testing-library/react';
import App from '../App';
import { MemoryRouter, Routes, Route } from 'react-router-dom';

it('renders', () => {
  const { container } = render(
    <MemoryRouter>
      <App />
    </MemoryRouter>,
  );
  expect(container).toMatchSnapshot();
});

describe('Navbar', () => {
  it('contains header on load', () => {
    render(
      <MemoryRouter>
        <App />
      </MemoryRouter>,
    );
    expect(screen.getByRole('banner')).toBeInTheDocument();
  });

  it('changes the page when a link is clicked', async () => {
    const user = userEvent.setup();

    render(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route path="/" element={<App />}>
            <Route index element={<h1>Home</h1>} /> {/* Home page */}
            <Route path="shop" element={<h1>Shop</h1>} /> {/* Shop page */}
          </Route>
        </Routes>
      </MemoryRouter>,
    );

    (expect(screen.getByRole('heading', { name: /home/i })).toBeInTheDocument(),
      await user.click(screen.getByRole('link', { name: /shop/i })));
    expect(screen.getByRole('heading', { name: /shop/i }));
  });

  it('has all the working links', async () => {
    const user = userEvent.setup();

    render(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route path="/" element={<App />}>
            <Route index element={<h1>Home</h1>} /> {/* Home page */}
            <Route path="shop" element={<h1>Shop</h1>} /> {/* Shop page */}
            <Route path="cart" element={<h1>Cart</h1>} /> {/* Shop page */}
          </Route>
        </Routes>
      </MemoryRouter>,
    );

    await user.click(screen.getByRole('link', { name: /shop/i }));
    expect(screen.getByRole('heading', { name: /shop/i })).toBeInTheDocument();
    await user.click(screen.getByRole('link', { name: /cart/i }));
    expect(screen.getByRole('heading', { name: /cart/i })).toBeInTheDocument();
    await user.click(screen.getByRole('link', { name: /home/i }));
    expect(screen.getByRole('heading', { name: /home/i })).toBeInTheDocument();
  });
});
