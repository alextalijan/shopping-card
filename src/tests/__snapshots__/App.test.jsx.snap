// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders 1`] = `
<div>
  <header
    class="header"
  >
    <a
      class="nav-link"
      data-discover="true"
      href="/"
    >
      Home
    </a>
    <a
      class="nav-link"
      data-discover="true"
      href="/shop"
    >
      Shop
    </a>
    <a
      class="nav-link"
      data-discover="true"
      href="/cart"
    >
      Cart 
      <span>
        (
        0
        )
      </span>
    </a>
  </header>
  <main />
</div>
`;
